### DISCUSSION TITLE
ESP32-S3 and SPI ILI9488 Causing Core Panic with All Examples

### DISCUSSION THREAD

**Discussion by GeorgeIoak** [Bodmer/TFT_eSPI#3392] [Compatible displays and setup files]
I tried everything I can think but I could not get the library to work. To confirm that the board and screen are working I configured an Arduino Sketch with LovyanGFX and the board worked on the first try.

I have Cored Debug Messages et to Trace so I can see that the correct GPIO pins are being redefined. I was trying to use the Read_User_Setup example to confirm the configuration but that also crashes so I relied on the Debug message to confirm that the correct setup is being read.

I'm using this setup:

```

#define USER_SETUP_ID 21  // Not sure this is needed

#define ILI9488_DRIVER

//#define TFT_INVERSION_OFF

#define TFT_MISO -1//13 // (leave TFT SDO disconnected if other SPI devices share MISO)
#define TFT_MOSI 11
#define TFT_SCLK 12
#define TFT_CS   10  // Chip select control pin
#define TFT_DC   18  // Data Command control pin
#define TFT_RST  -1//39  // Reset pin (could connect to RST pin)

#define TFT_BL   12            // LED back-light control pin
#define TFT_BACKLIGHT_ON HIGH  // Level to turn ON back-light (HIGH or LOW)

#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:.
#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts

#define SMOOTH_FONT


//#define SPI_FREQUENCY  20000000
// #define SPI_FREQUENCY  27000000
#define SPI_FREQUENCY  40000000
// #define SPI_FREQUENCY  80000000

// Optional reduced SPI frequency for reading TFT
#define SPI_READ_FREQUENCY  16000000

#define SPI_TOUCH_FREQUENCY  2500000
```

But as you can see I experimented with disabling TFT_MISO, TFT_RST, and SPI_FREQUENCY (most tests ran at 27MHz)

I originally started with trying to use LVGL with TFT_eSPI but when all those attempts failed I switched to just checking TFT_eSPI alone. 

So since LovyanGFX works that confirms that my hardware is OK but I can't find anything to explain why my ESP32-S3 (no PSRAM) causes panics.

**Comment by ChassaB** [comment]
I’m having the same problem. Tested with ILI9488 and ILI92225. Always panics as soon as tft.init
 Both screens working fine on an esp32. Nothing else on  SPI or any thing on any pins. Just the tft.
The Esp32 S3 SPI works with an SD card. Don’t know where to go either.

**Reply by DovahBrownies** [reply]
@GeorgeIoak If you're on a newer board version than `2.0.14`, there seems to be a bug that causes the core panic. Try going back to that version. For example, in the (old) Arduino IDE:

![image](https://github.com/Bodmer/TFT_eSPI/assets/35106956/39ac877b-edb3-4588-a229-9a76992779ed)

**Comment by GeorgeIoak** [comment]
I thought it might have something to do with 3.x but since LovyanGFX works I ignored trying to downgrade the core. Is it 3.0 and above that don't seem to work with TFT_eSPI??

**Reply by ChassaB** [reply]
I'll give it a try too. Found that post yesterday as well but haven't tried the downgrade yet. Just used an ESP32 instead for what I was doing. Now I can try switching back to the S3

**Comment by GeorgeIoak** [comment]
OK, I downgraded to 1.0.17 and that didn't change anything so I did a little more researching and came across this in some of the user defines

```
// The ESP32 has 2 free SPI ports i.e. VSPI and HSPI, the VSPI is the default.
// If the VSPI port is in use and pins are not accessible (e.g. TTGO T-Beam)
// then uncomment the following line:
#define USE_HSPI_PORT
```

adding that define magically made not only the code work but I was also able to get LVGL examples to work too.

I dug a little into the code and saw that these defines change what SPI port is used and for the ESP32-S3 only SPI2/3 are available but the ESP32-S3 doesn't have HSPI/VSPI like the ESP32, it has FSPI (flexible SPI).

What confuses me a bit is that not all of the example ESP32-S3 User_Setups have this or the USE_FSPI_PORT defines but if the library defaults to SPI0/1 why aren't others who have ESP32-S3 having problems?

I'm doing some additional testing and after that I'll go back to 3.x Core and see if that breaks anything

**Reply by JimDrewGH** [reply]
> OK, I downgraded to 1.0.17 and that didn't change anything so I did a little more researching and came across this in some of the user defines

I am guessing you meant 2.0.17?

You need to downgrade to 2.0.14 in order for this problem to go away for the ESP32-S2/S3/C3/H6.

Please let us know if you find a solution that actually works for the S2/S3/C3/H6 versions with 2.0.15 or later of the core.

**Reply by DovahBrownies** [reply]
> adding that define magically made not only the code work but I was also able to get LVGL examples to work too.

Now sure how specifying the HSI port worked for you; Though I'm not an expert in this field.
I am running an ESP32-S3-DevkitC-1, and also using a **ILI9488** with capacitive touch. I didn't have to do that. Just had to downgrade the board firmware.

But if it's working, you should close this issue and mark an answer! :)

**Reply by GeorgeIoak** [reply]
Yes, sorry, `1.0.17` was a typo and meant to be `2.0.17`

Downgrading to 2.0.14 might very well work but that's not an answer to the issue, especially since I've stated that the LovyanGFX library works with 3.x.

It appears to be an issue as to how the SPI port numbers are assigned and I haven't had time to dig into this further. I opened this discussion so that it could be discussed and perhaps find the root cause and propose a fix that works for everyone that doesn't involve downgrading the Arduino Core.

**Reply by JimDrewGH** [reply]
I agree, downgrading is never a "solution", just a band-aide.  In the "issues" area there seems to be a fix that involves re-defining the SPI port(s).

https://github.com/Bodmer/TFT_eSPI/issues/3384#issuecomment-2211470238

**Reply by AlexBestoso** [reply]
Can confirm that there appears to be an issue with how the hardware and virtual spi interfaces play with the rest of the system.

Initializing using HSPI does get the SPI interface to function. 

Though I've been experiencing a problem where after I initialize a slave device, the device breaks after initializing a second SPI slave.

IE: LVGL using eSPI works fine. 
But if I then initialize RadioLib or my SD card, I get a hang on `tft_Write_8 TFT_WRITE_BITS(C, 8)` 
Something to do with `SPI_USR` and  `_spi_cmd` As these appear to be what's causing the offending while loop to get stuck.

I gotta dig deeper to see why these fail after secondary slave initialization (slightly different problem, but I believe the issues are related.)
